File: src/main/java/io/github/chaosdave34/kitpvp/kits/impl/kits/MagicianKit.java
package io.github.chaosdave34.kitpvp.kits.impl.kits;

import io.github.chaosdave34.kitpvp.ExtendedPlayer;
import io.github.chaosdave34.kitpvp.items.CustomItemHandler;
import io.github.chaosdave34.kitpvp.kits.Kit;
import org.bukkit.Color;
import org.bukkit.Material;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.player.PlayerItemHeldEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.PotionMeta;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;
import org.bukkit.potion.PotionType;

public class MagicianKit extends Kit {
    public MagicianKit() {
        super("magician", "Magician");
    }

    @Override
    public ItemStack getHeadContent() {
        ItemStack leatherHelmet = new ItemStack(Material.LEATHER_HELMET);
        leatherHelmet.addEnchantment(Enchantment.PROTECTION_ENVIRONMENTAL, 2);
        setLeatherArmorColor(leatherHelmet, Color.RED);
        return leatherHelmet;
    }

    @Override
    public ItemStack getChestContent() {
        ItemStack leatherChestplate = new ItemStack(Material.LEATHER_CHESTPLATE);
        leatherChestplate.addEnchantment(Enchantment.PROTECTION_ENVIRONMENTAL, 2);
        setLeatherArmorColor(leatherChestplate, Color.NAVY);
        return leatherChestplate;
    }

    @Override
    public ItemStack getLegsContent() {
        ItemStack leatherLeggings = new ItemStack(Material.LEATHER_LEGGINGS);
        leatherLeggings.addEnchantment(Enchantment.PROTECTION_ENVIRONMENTAL, 2);
        setLeatherArmorColor(leatherLeggings, Color.NAVY);
        return leatherLeggings;
    }

    @Override
    public ItemStack getFeetContent() {
        ItemStack leatherBoots = new ItemStack(Material.LEATHER_BOOTS);
        leatherBoots.addEnchantment(Enchantment.PROTECTION_ENVIRONMENTAL, 2);
        setLeatherArmorColor(leatherBoots, Color.BLACK);
        return leatherBoots;
    }

    @Override
    public ItemStack[] getInventoryContent() {
        return new ItemStack[]{
                CustomItemHandler.MAGIC_WAND.build(),
                new ItemStack(Material.BOOK),
                new ItemStack(Material.WATER_BUCKET),
        };
    }

    @Override
    public ItemStack[] getKillRewards() {
        ItemStack splashPotion = new ItemStack(Material.SPLASH_POTION, 2);
        splashPotion.editMeta(PotionMeta.class, potionMeta -> potionMeta.setBasePotionType(PotionType.INSTANT_HEAL));

        return new ItemStack[]{
                splashPotion,
                new ItemStack(Material.GILDED_BLACKSTONE, 32),
        };
    }

    @EventHandler
    public void onSlotChange(PlayerItemHeldEvent e) {
        Player p = e.getPlayer();
        ExtendedPlayer extendedPlayer = ExtendedPlayer.from(p);
        if (extendedPlayer.getSelectedKitId().equals(getId()) && extendedPlayer.getGameState() == ExtendedPlayer.GameState.KITS_IN_GAME) {
            ItemStack slot = p.getInventory().getItem(e.getNewSlot());

            // Todo: improve this
            if (slot == null || slot.getType() != Material.BOOK) {
                p.removePotionEffect(PotionEffectType.SPEED);
                p.removePotionEffect(PotionEffectType.INVISIBILITY);
                p.getInventory().setHelmet(getHeadContent());
                p.getInventory().setChestplate(getChestContent());
                p.getInventory().setLeggings(getLegsContent());
                p.getInventory().setBoots(getFeetContent());
            } else {
                p.addPotionEffect(new PotionEffect(PotionEffectType.SPEED, -1, 3, false, false, false));
                p.addPotionEffect(new PotionEffect(PotionEffectType.INVISIBILITY, -1, 0, false, false, false));
                p.getInventory().setHelmet(new ItemStack(Material.AIR));
                p.getInventory().setChestplate(new ItemStack(Material.AIR));
                p.getInventory().setLeggings(new ItemStack(Material.AIR));
                p.getInventory().setBoots(new ItemStack(Material.AIR));
            }
        }
    }
}

File: src/main/kotlin/io/github/chaosdave34/kitpvp/kits/impl/kits/MagicianKit.kt
package io.github.chaosdave34.kitpvp.kits.impl.kits

import io.github.chaosdave34.kitpvp.items.CustomItemHandler
import io.github.chaosdave34.kitpvp.kits.Kit
import io.github.chaosdave34.kitpvp.ultimates.Ultimate
import io.github.chaosdave34.kitpvp.ultimates.UltimateHandler
import org.bukkit.Color
import org.bukkit.Material
import org.bukkit.enchantments.Enchantment
import org.bukkit.event.EventHandler
import org.bukkit.event.player.PlayerItemHeldEvent
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.PotionMeta
import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType
import org.bukkit.potion.PotionType

class MagicianKit : Kit("magician", "Magician") {

    override fun getHeadContent(): ItemStack {
        val helmet = ItemStack.of(Material.LEATHER_HELMET)
        helmet.addEnchantment(Enchantment.PROTECTION, 2)
        helmet.setLeatherArmorColor(Color.RED)
        return helmet
    }

    override fun getChestContent(): ItemStack {
        val chestplate = ItemStack.of(Material.LEATHER_CHESTPLATE)
        chestplate.addEnchantment(Enchantment.PROTECTION, 2)
        chestplate.setLeatherArmorColor(Color.NAVY)
        return chestplate
    }

    override fun getLegsContent(): ItemStack {
        val leggings = ItemStack.of(Material.LEATHER_LEGGINGS)
        leggings.addEnchantment(Enchantment.PROTECTION, 2)
        leggings.setLeatherArmorColor(Color.NAVY)
        return leggings
    }

    override fun getFeetContent(): ItemStack {
        val boots = ItemStack.of(Material.LEATHER_BOOTS)
        boots.addEnchantment(Enchantment.PROTECTION, 2)
        boots.setLeatherArmorColor(Color.BLACK)
        return boots
    }

    override fun getInventoryContent(): Array<ItemStack?> {
        return arrayOf(
            CustomItemHandler.MAGIC_WAND.build(),
            ItemStack.of(Material.BOOK),
            ItemStack.of(Material.WATER_BUCKET),
        )
    }

    override fun getKillRewards(): Array<ItemStack> {
        val splashPotion = ItemStack.of(Material.SPLASH_POTION, 2)
        splashPotion.editMeta(PotionMeta::class.java) { potionMeta: PotionMeta -> potionMeta.basePotionType = PotionType.HEALING }

        return arrayOf(
            splashPotion,
            ItemStack.of(Material.GILDED_BLACKSTONE, 32),
        )
    }

    override fun getUltimate(): Ultimate = UltimateHandler.LEVITATE

    @EventHandler
    fun onSlotChange(event: PlayerItemHeldEvent) {
        val player = event.player
        if (player.isKitActive()) {
            val slot: ItemStack? = player.inventory.getItem(event.newSlot)

            val inventory = player.inventory

            if (slot?.type != Material.BOOK) {
                player.removePotionEffect(PotionEffectType.SPEED)
                player.removePotionEffect(PotionEffectType.INVISIBILITY)
                inventory.helmet = getHeadContent()
                inventory.chestplate = getChestContent()
                inventory.leggings = getLegsContent()
                inventory.boots = getFeetContent()
            } else {
                player.addPotionEffect(PotionEffect(PotionEffectType.SPEED, -1, 3, false, false, false))
                player.addPotionEffect(PotionEffect(PotionEffectType.INVISIBILITY, -1, 0, false, false, false))
                inventory.helmet = ItemStack.of(Material.AIR)
                inventory.chestplate = ItemStack.of(Material.AIR)
                inventory.leggings = ItemStack.of(Material.AIR)
                inventory.boots = ItemStack.of(Material.AIR)
            }
        }
    }
}